sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala

```diff
@@ -60,980 +79,305 @@ case class ScalaUDF(
+          func(convertedRow.get(0), convertedRow.get(1), convertedRow.get(2), convertedRow.get(3),
+            convertedRow.get(4), convertedRow.get(5), convertedRow.get(6), convertedRow.get(7),
+            convertedRow.get(8), convertedRow.get(9), convertedRow.get(10), convertedRow.get(11),
+            convertedRow.get(12), convertedRow.get(13), convertedRow.get(14), convertedRow.get(15),
+            convertedRow.get(16), convertedRow.get(17), convertedRow.get(18), convertedRow.get(19),
+            convertedRow.get(20), convertedRow.get(21))
+        }
+    }
   }
 
   // scalastyle:on
 
-  // Generate codes used to convert the arguments to Scala type for user-defined funtions
-  private[this] def genCodeForConverter(ctx: CodeGenContext, index: Int): String = {
-    val converterClassName = classOf[Any => Any].getName
-    val typeConvertersClassName = CatalystTypeConverters.getClass.getName + ".MODULE$"
-    val expressionClassName = classOf[Expression].getName
-    val scalaUDFClassName = classOf[ScalaUDF].getName
-
-    val converterTerm = ctx.freshName("converter")
-    val expressionIdx = ctx.references.size - 1
-    ctx.addMutableState(converterClassName, converterTerm,
-      s"this.$converterTerm = ($converterClassName)$typeConvertersClassName" +
-        s".createToScalaConverter(((${expressionClassName})((($scalaUDFClassName)" +
-          s"expressions[$expressionIdx]).getChildren().apply($index))).dataType());")
-    converterTerm
-  }
-
   override def genCode(
       ctx: CodeGenContext,
       ev: GeneratedExpressionCode): String = {
 
     ctx.references += this
+    val scalaUDFTermIdx = ctx.references.size - 1
 
     val scalaUDFClassName = classOf[ScalaUDF].getName
     val converterClassName = classOf[Any => Any].getName
     val typeConvertersClassName = CatalystTypeConverters.getClass.getName + ".MODULE$"
     val expressionClassName = classOf[Expression].getName
-
-    // Generate codes used to convert the returned value of user-defined functions to Catalyst type
-    val catalystConverterTerm = ctx.freshName("catalystConverter")
-    val catalystConverterTermIdx = ctx.references.size - 1
-    ctx.addMutableState(converterClassName, catalystConverterTerm,
-      s"this.$catalystConverterTerm = ($converterClassName)$typeConvertersClassName" +
-        s".createToCatalystConverter((($scalaUDFClassName)expressions" +
-          s"[$catalystConverterTermIdx]).dataType());")
+    val expressionEncoderClassName = classOf[ExpressionEncoder[Row]].getName
+    val rowEncoderClassName = RowEncoder.getClass.getName + ".MODULE$"
+    val structTypeClassName = StructType.getClass.getName + ".MODULE$"
+    val rowClassName = Row.getClass.getName + ".MODULE$"
+    val rowClass = classOf[Row].getName
+    val internalRowClassName = classOf[InternalRow].getName
+    // scalastyle:off
+    val javaConversionClassName = scala.collection.JavaConversions.getClass.getName + ".MODULE$"
```

Comment:
`JavaConversions` has been banned for implicit conversion between Java and Scala types. However, we are not going to use it in Scala side but use it in generated Java codes. `JavaConverters` doesn't provide simple and direct method to call for the purpose here. So we turn off scalastyle here temporarily.